Distrust certs issued after 00:00:00 Oct. 21, 2016 by StartCom and WoSign

shigeki opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
crypto, tls and test.

##### Description of change
<!-- Provide a description of the change below this comment. -->
Per discussion of #9434, this PR checks the certs issued by StartCom and WoSign and if notBefore is after 00:00:00 Oct. 21 2016, the tls client connection is failed with CERT_REVOKED error.

This also includes CNNIC whitelist update since #1895 that expired certs were removed and a minor bug fix of `test/parallel/test-tls-cnnic-whitelist.js` which came from 2bc7841d0fcdd066fe477873229125b6f003b693.

R: @bnoordhuis

shigeki commented:
CI is running on https://ci.nodejs.org/job/node-test-pull-request/4785/.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2678,9 +2682,40 @@ inline X509* FindRoot(STACK_OF(X509)* sk) {
+  const unsigned char* startcom_wosign_data;
+  X509_NAME* startcom_wosign_name;
+
+  for (size_t i = 0; i < arraysize(StartComAndWoSignDNs); i++) {
```
You could write this as `for (auto it : StartComAndWoSignDNs) {`

shigeki replied:
Done. I used `const auto&` since it is just a reference.

bnoordhuis commented on test/fixtures/keys/Makefile:
```diff
@@ -58,6 +58,17 @@ fake-cnnic-root-cert.pem: fake-cnnic-root.cnf fake-cnnic-root-key.pem
+	openssl genrsa -out fake-startcom-root-key.pem 2048
+
+fake-startcom-root-cert.pem: fake-startcom-root.cnf fake-startcom-root-key.pem
+	openssl req -new -x509 -days 9999 -config fake-startcom-root.cnf -key fake-startcom-root-key.pem -out fake-startcom-root-cert.pem
```
Can you wrap this at 80 columns?

shigeki replied:
Done. I missed `Makefile` is out of lint checks.

not-an-aardvark commented on src/StartComAndWoSignData.inc:
```diff
@@ -0,0 +1,89 @@
+  0x31, 0x2C, 0x30, 0x2A, 0x06, 0x03, 0x55, 0x04, 0x03, 0x13, 0x23, 0x53, 0x74,
+  0x61, 0x72, 0x74, 0x43, 0x6F, 0x6D, 0x20, 0x43, 0x65, 0x72, 0x74, 0x69, 0x66,
+  0x69, 0x63, 0x61, 0x74, 0x69, 0x6F, 0x6E, 0x20, 0x41, 0x75, 0x74, 0x68, 0x6F,
+  0x72, 0x69, 0x74, 0x79, 0x20, 0x47, 0x32,
```
Just as a precaution: How did you compute these arrays? (I'm hoping to avoid an issue like [this](http://www.openwall.com/lists/oss-security/2016/02/01/4).)

jasnell commented on src/node_crypto.cc:
```diff
@@ -2678,9 +2682,40 @@ inline X509* FindRoot(STACK_OF(X509)* sk) {
 
 
-// Whitelist check for certs issued by CNNIC. See
+inline bool CertIsStartComOrWoSign(X509_NAME* name) {
```
Hmm... would there possibly be a more generic way of doing this? Adding functions that are specific to StartCom and WoSign seems... troubling.

shigeki replied:
I don't think it leads troubling because this function checks the cert subject name with a specific type of structure provided by Mozilla only for StartCom and WoSign distrusting.
This function (and this features) will be removed after deprecating their root cert when all issued certs are expired. So I don't think it need to be more generic.

jasnell commented on test/parallel/test-tls-startcom-wosign-whitelist.js:
```diff
@@ -0,0 +1,91 @@
+'use strict';
+var common = require('../common');
```
`const`?

shigeki replied:
Done.

shigeki commented:
Fixed this according to the comments except the one from @jasnell . 

@indutny Sorry for missing you to mention. Please look at this if you have time.

indutny commented on src/node_crypto.cc:
```diff
@@ -2678,9 +2682,40 @@ inline X509* FindRoot(STACK_OF(X509)* sk) {
+  if (!CertIsStartComOrWoSign(root_name))
+    return true;
+
+  time_t october_21_2016 = static_cast<time_t>(1477008000);
```
could it be `const`?

shigeki replied:
It's not. I'm not sure why.
```c
int X509_cmp_time(const ASN1_TIME *s, time_t *t);
````

shigeki commented:
CI is again running on https://ci.nodejs.org/job/node-test-pull-request/4795/ .

seishun commented on src/node_crypto.cc:
```diff
@@ -2678,9 +2682,40 @@ inline X509* FindRoot(STACK_OF(X509)* sk) {
+  const unsigned char* startcom_wosign_data;
+  X509_NAME* startcom_wosign_name;
+
+  for (const auto& it : StartComAndWoSignDNs) {
```
Minor, but calling it `it` is confusing, as it's not an iterator.

shigeki replied:
@seishun Yes, I rename `it` to `dn` after 'distinguished name`.  Thanks.

ChALkeR commented:
@bnoordhuis, does that look good to you now?

shigeki commented:
I've just resolved the conflict of `test/parallel/test-tls-cnnic-whitelist.js` and rebased against the latest master.

Firefox51 is to be stable on Jan 24th and Chrome56 is on Jan 31st, 2017. They will begin to have  WoSing/StartCom checks in their stable release.

I think it is better to land this in the end of Jan, 2017 to align the browser release.

shigeki commented:
Chrome 56 is released today. It is ready to land this. The approval from @bnoordhuis is needed. How is it?

shigeki commented:
@bnoordhuis Oh, typo. Thanks. Fixed.

I also added an announcement link from apple https://support.apple.com/en-us/HT204132 in the commit log. I will land this tomorrow after rebasing and running CI.

After that, it should be discussed if it should be backported to LTS.

jasnell commented:
I'm +1 on backporting this to 6 and 4. @nodejs/lts

shigeki commented:
Running CI with rebasing in the current master for final check before landing. https://ci.nodejs.org/job/node-test-pull-request/6116/

shigeki commented:
There were lint errors for using var and assert.equal. The fixes were made in 00017cef833274b41d6c0c7bc63e4bce10e8c456. 
There are still errors on CI on windows due to git errors. Wait for fixing for now.

joaocgreis commented:
Aborted CI, only `arm-fanned` was pending because of https://github.com/nodejs/build/issues/611

shigeki commented:
@joaocgreis Thanks for notifying. I'm going to submit a new CI job, hoping working fine.

shigeki commented:
CI of https://ci.nodejs.org/job/node-test-commit/7652/ is all green. Landed in 89217d145097590e37d11db6e98c317d154be689 and d4464241d7ea84e6c9c2cd9caed92cf6ed8743b4. Thanks for reviewing.

I also put labels of lts-watch for v4 and v6. Please change them if anyone has any objections.

sam-github commented:
@nodejs/lts this is semver-major, but is allowed to backport because it is a security fix, do we have a label for these kinds of PRs?
